### PR TITLE
fix(demo-reel): verify cast provenance instead of trusting path existence (#808)

### LINKED ISSUE #808: demo-reel adapter: verify cast provenance instead of trusting path existence
> **BLOCKED — do not start this until PR #805 is merged.**
>
> This issue changes the demo-reel adapter and engine. PR #805 also modifies those (`.claude/skills/demo-reel-adapter/SKILL.md`, `.claude/skills/demo-reel/reel.sh`, `upload.sh`) for the private-by-default upload. The overlap is smaller than #806's or #810's but still a conflict.
>
> Check first: `gh pr view 805 --json state,mergedAt`. If it is not `MERGED`, stop and pick something else.
>
> Priority note: #805 made the reel upload **private by default**, so a stale or unredacted cast is no longer automatically third-party-visible. That bounds this issue's urgency — take #806, #810 and #811 before it.

---

## Problem

`.claude/skills/demo-reel-adapter/build.sh` decides a cast is publishable from **path existence** plus two static facts about the test: that it is branch-scoped and that its `tests/CATALOG.md` entry carries the ` [reel]` marker (`build.sh:288-320`, `:347-375`). It records no **run identity**, no **commit**, no **redaction version**, no **timestamp requirement** and no proof that the test it came from actually **passed** in the current run. The reel engine (`.claude/skills/demo-reel`) performs no credential redaction of its own — redaction happens only inside the harness, when the cast is written.

So the adapter cannot distinguish a cast this run produced from one an older revision left behind. That matters because a reel-eligible clip is by definition a **real-agent** recording (CLAUDE.md rule 4) and its URL goes into the PR body *and* the changelog fragment, which flows into the **public release notes**. The upload itself is **private by default** since PR #805 (`.claude/skills/demo-reel/upload.sh`), so a stale clip is not third-party-visible until a human deliberately flips the video — that bounds the blast radius, it does not make publishing a cast from an unknown revision correct.

## What PR #805 already fixed, stated narrowly

The second security audit of PR #805 found the first version of this section **overstated** the fix, so here is the accurate version. Two discards landed, and each covers exactly the routes that reach its own call site:

- `TuiDeck::try_launch_inner` calls `discard_previous_recording` **before it spawns anything**, so once a test reaches the launch, a `SIGKILL`, a nextest timeout, a Ctrl-C or a failed re-recording cannot leave the previous run's artifacts in place.
- `_skip_if_err` (the helper behind `skip_unless!`) does the same before either of its two outcomes, so a runtime skip — the worst shape, because nextest reports it as PASSED — cannot either.
- A deletion that FAILS for any reason other than `NotFound` now **panics** rather than warning and continuing, so "the discard ran but could not remove the file" is no longer a silent route to a stale artifact.

`tests/harness_isolation.rs` carries all three guards.

## What is left, and why clearing does not close it

Clearing is keyed to reaching one of those two call sites. Two routes do not:

1. **A filtered run** — which is the normal way anyone works, and what CLAUDE.md rule 5 now asks for — does not select every reel-marked test, so an unselected test's older cast is never touched and still passes the adapter's `test -f`.
2. **The preflight runs BEFORE the skip-path discard.** `skip_unless!(check_claude_available())` evaluates the check first and only then hands its `Result` to `_skip_if_err`, so a kill, a timeout or an abort inside a `check_*_available` — or inside an importer it calls, which is where the credential work happens — happens before anything is discarded.

In neither case can the pipeline tell that a clip came from a different revision than the rest of the reel, possibly one predating the redaction fixes in #785/#502.

## Proposal

Make the adapter verify **provenance** rather than existence:

- The harness writes a small sidecar beside each dump (or a header field in the `.cast`) recording at least: the run's identity, `git describe`/commit of the binary under test, the redaction implementation version, the wall-clock time, and the test's **outcome**.
- `build.sh` refuses a cast whose sidecar is absent, whose commit does not match the reel build's `HEAD`, or whose outcome is not `passed`, and says which clip it dropped and why rather than silently including or excluding it.

Provenance is what covers both remaining routes at once, which is why the answer is not a third discard.

Three caveats to keep on record, all from the #805 audits:

- A **passing** test lowers the probability of an error dump but does not change the boundary — successful terminal output can still contain a credential, so provenance is not a substitute for the harness redaction.
- The harness redaction is itself **best-effort**: a blocklist that can only remove registered values, with two reproduced gaps tracked in #810. Provenance and redaction do not cover for each other.
- Watching the finished reel before publishing is useful defence in depth, **not** a reliable secret scanner.

## Scope

Adapter and harness only; no change to the redaction matcher, which #810 covers. `.claude/skills/demo-reel-adapter/tests/adapter_test.sh` is the existing test surface (run via `Taskfile.yml`) and is where the new refusals belong.

Refs: #502, #785, #810, PR #805.

### PR DESCRIPTION
**Pull request by vfarcic**
Closes #808.

## The residual, not what #805 already fixed

The adapter decided a cast was publishable from **path existence** plus two static facts about the *test* — that its source changed on the branch, and that its `tests/CATALOG.md` line carries the ` [reel]` marker (`build.sh:288-320`, `:347-375` before this change). None of those is a fact about the **artifact**, so a `full-stream.cast` an older revision left on disk satisfied every one of them.

PR #805's three guards already landed and are untouched here: `try_launch_inner` discards before it spawns anything, `_skip_if_err` discards before either outcome, and a deletion failing for anything but `NotFound` panics. Those close every route that **reaches** one of those two call sites. Two do not, and both want provenance rather than a third discard:

1. **A filtered run** — the normal way anyone works, and what CLAUDE.md rule 5 asks for — never selects the test at all, so nothing on the discard path executes and an unselected test's older cast is untouched.
2. **`skip_unless!` evaluates its preflight before `_skip_if_err` is entered.** `skip_unless!(check_claude_available())` calls the check first and only then hands the `Result` over, so a kill or abort inside a `check_*_available` — or inside an importer it calls, which is where the credential work happens — lands before the skip-path discard.

## What is recorded, and where

`tests/common/mod.rs` gains `write_provenance`, called last from `dump_recordings`, writing `provenance.json` beside the cast:

| Field | Source | Adapter |
| --- | --- | --- |
| `schema` | `RECORDING_PROVENANCE_SCHEMA` | **refuses** an unknown value |
| `outcome` | `std::thread::panicking()` at `Drop` | **refuses** anything but `passed` |
| `commit` | short sha parsed from the binary's compile-time `DAD_BUILD_ID` | **refuses** a value that is not the reel build's `HEAD` |
| `dirty` | `DAD_BUILD_ID`'s `-dirty` suffix | reported loudly, refuses nothing |
| `build_id` | `env!("DAD_BUILD_ID")` | reported |
| `run_id` | `NEXTEST_RUN_ID` | reported; **differing ids across clips only note the count** |
| `redaction_version` | `RECORDING_REDACTION_VERSION` | reported; required present |
| `recorded_at_unix` | `SystemTime::now()` | reported, never thresholded |

`commit` comes from the **binary**, not from a `git rev-parse` at dump time. A tree-time read would name the revision the run was *launched* against, which a stale binary makes a lie, and there would then be two commit facts that can disagree. One fact, sourced from the artifact's own producer.

## Which residual route each fact closes

- **`commit` closes both.** A cast an older revision left behind — whether because a filtered run never selected the test, or because a kill inside a preflight preceded the discard — names that older revision, and is refused.
- **`outcome` closes a large slice of both, independently.** `Drop` dumps *because* the test panicked, so a failure diagnostic is the commonest artifact on disk; it is refused whatever its commit.
- **Sidecar-absent closes the pre-provenance case and a torn dump.** Written last of the dump's artifacts, so a dump that dies partway leaves a cast the adapter refuses. `provenance.json` is also **first** in `RECORDING_ARTIFACTS`, so a discard that panics partway has already removed the sidecar that would have vouched for whatever survives.
- **`run_id` closes neither on its own, and is not claimed to.** It is an audit-trail fact and a cross-clip consistency signal. It cannot be a gate: a reel legitimately assembles clips from several filtered runs at one commit, which is precisely the working pattern rule 5 asks for.

## What it deliberately does not cover

- **Not recency.** A passing clip recorded at this commit by an earlier run is accepted, and correctly so — the code that produced it is the code under test. Provenance establishes revision and outcome, not "this run".
- **Under `dirty`, not even the revision.** One commit then covers more than one working state, so a clip from an earlier dirty state at the same commit is accepted. That is why the flag is surfaced loudly rather than swallowed, and why it does not refuse: recording from a dirty tree is the ordinary dogfood case.
- **`outcome: passed` is narrow.** It means the harness was not unwinding a panic when the deck was dropped — not that nextest reported the test as passed. A test that panics after its deck is dropped is not observed; a test killed outright reaches no `Drop`, writes no sidecar, and is covered by #805's launch-time discard instead.
- **Nothing about content.** The harness redaction is a best-effort blocklist with two reproduced gaps (#810). Provenance and redaction do not cover for each other, and a passing test lowers the probability of a credential-bearing dump without changing the boundary.
- **Not a signature.** The sidecar sits in the same gitignored directory as the cast, so whoever can write one can write the other.
- **Not the upload's visibility.** Untouched; private-by-default from #805 stands.

## Where it is enforced

**Assembly only**, and that is sufficient rather than a shortcut: every route to a manifest runs `assemble` — the `reel` pipeline and the standalone `build.sh assemble <id...>` that an injected id list would use. `select_ids` is unchanged, because "in scope" is a statement about the test while provenance is a statement about the artifact; `build.sh select` therefore stays a pure scope query with no git-HEAD dependency of its own.

The gate is **per-clip**: one stale sidecar drops one clip and the reel is still built from what survived, with the omission named where the manifest is announced. Dropping a whole reel over one stale clip would only teach people to bypass the gate. `skip_message` gained a third bucket so a stale-only skip cannot borrow the marker wording — blaming the marker for a stale cast is the issue #735 misattribution in a new bucket.

## Tests added

**Eleven cases in `.claude/skills/demo-reel-adapter/tests/adapter_test.sh`** (section `(v)`), each mutating one field of a copy of the fixture tree: absent sidecar `(v-a)`, different commit `(v-b)`, non-passing outcome `(v-c)`, unknown schema `(v-d)`, corrupt JSON `(v-e)`, missing required field `(v-e2)`, one-stale-of-two publishes the other `(v-f)`, dirty publishes with a warning `(v-g)`, two run ids publish with a note `(v-h)`, no resolvable commit **dies** rather than publishing unchecked `(v-i)`, and with the override unset the expected commit is resolved from the real git `HEAD` `(v-j)`.

**Three fast-tier guards in `tests/harness_isolation.rs`:**
- `the_build_id_parse_yields_a_commit_only_when_one_is_really_there` — every `DAD_BUILD_ID` shape `build.rs` composes plus injected ones, including the `-gamma.1-g<sha>` prerelease case that makes the *last* `-g` load-bearing.
- `the_provenance_sidecar_is_written_last_and_discarded_first` — both orders, both fail-closed, pointing opposite ways.
- `the_provenance_contract_the_adapter_reads_is_the_one_the_harness_writes` — the shell adapter's jq read set against the Rust harness's `json!` write set, plus schema-number agreement. The two ends are in different languages and neither compiles the other, so drift would refuse every clip *silently* until somebody next tried to publish.

`(v-e2)` earned its place immediately: it caught a real bug in my first implementation, where `read` with a tab `IFS` collapsed consecutive delimiters (tab is IFS *whitespace* in bash), shifting every field after an empty one left and admitting a sidecar with no commit. Now joined on US (0x1f).

## Verification

| Command | Result |
| --- | --- |
| `cargo fmt --check` | clean |
| `cargo clippy --workspace --all-targets --features e2e,e2e-live -- -D warnings` | clean (all four flags) |
| `cargo test-fast` | **4018/4018** |
| `cargo nextest run --workspace --test harness_isolation` | 115/115, incl. the three new guards |
| `task reel-adapter-test` | 18 sections, all pass |
| `cargo xtask` (linkage-check) | ok, 11 rules |

**`write_provenance` was executed for real, not only source-guarded.** `DOT_AGENT_DECK_RECORD=1 cargo test-e2e pane_input_021` (lane 1, no credential) wrote a valid sidecar on both a failing and a passing run — `outcome` flipping `failed` → `passed`, `run_id` matching nextest's printed run id, and the previous run's sidecar correctly replaced. Feeding that **unmodified** sidecar to the adapter with **no commit override** included it against the real `HEAD` (`364a182` prefix-matching `364a182cfb43…`, with the dirty warning firing), while a sibling recording from another revision was refused and named.

One pre-existing flake seen and dismissed per CLAUDE.md rule 6: `delegate_022_delegate_during_an_in_flight_close_brings_the_role_back` failed once in a 4018-test parallel run and passed alone in 4.3s — the parallel-contention pattern rule 5 documents from #415. It is not a `TuiDeck` test and touches no path this PR changes; the re-run of the full tier was 4018/4018.

## Notes

- **Changelog:** `changelog.d/808.misc.md`. This is internal tooling that never reaches a user's machine, but the repo writes `misc` fragments for exactly this class (#769 workflows, #815 devbox scripts, #834 xtask sandbox), and there is an operator-visible consequence worth recording — a reel build can now refuse everything, and the remedy is to re-record rather than to reach for the override.
- **Engine untouched.** Provenance is repo-specific, so the `{title, description, clip}` manifest contract with `.claude/skills/demo-reel/` stays exactly as it was. I confirmed no `Skill [skip ci]` sync commit has ever touched `.claude/skills/demo-reel/`, but it needed no edit.
- **No cross-version contract check needed** (rule 12): no daemon, protocol, orchestration or hook path is touched. `provenance.json` is a new file in a gitignored dev-time directory read only by a local shell script, not a cross-process compatibility contract.
- **`e2e-deterministic`** is red on `main` and every branch from #818; PR #837 fixes the remit half. It is not a required check.
- **Docs:** the adapter `SKILL.md` gains a `Cast provenance` section with the field table and the boundary; `docs/develop/demo-reel.md` gains a shorter version pointing at it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/skills/demo-reel-adapter/SKILL.md [hunk 1]
```diff
@@ -185,9 +234,16 @@ A **re-runnable, pure-shell** test (no `agg`/`ffmpeg`, no git, no network — so
...
-   invokes the engine (a stub) with no near-miss reported.
-
-Sections 1–2 are pure shell. Section 3 exercises the selection half, which exists to run `git diff`, so it shells out to **git** — building every repository inside its own `mktemp -d` with the ambient git configuration switched off, so it can neither read nor write the checkout it runs in, and with no network and no sleep (the same discipline CLAUDE.md rule 5 sets for the `xtask` real-git tests). It **skips**, without failing, where `git` is unavailable.
+   invokes the engine (a stub) with no near-miss reported;
+4. the **cast-provenance** gate refuses a clip whose sidecar is absent, corrupt,
+   of an unknown schema, missing a required field, recorded at a different
+   commit, or recorded from a non-passing run — while a valid sidecar publishes,
+   one stale sidecar drops **one** clip rather than the reel, a dirty-tree or
+   multi-run reel publishes **with a warning** rather than a refusal, and an
+   adapter that cannot resolve any commit to check against **dies** instead of
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply .claude/skills/demo-reel-adapter/SKILL.md:243]: Checked each of the three files named, and declining this one — with the measurements, since two of the three claims do not hold.

**`changelog.d/808.misc.md`** — no wrapped paragraph in it. Every prose line is a whole paragraph: `awk '{print NR": "length($0)}'` gives 76, 675, 765, 987, 369, 1110, 1047 with blank lines between. Nothing to unwrap.

**`docs/develop/demo-reel.md`** — same. The five added prose lines measure 595, 630, 914 and 443 characters (plus the 58-character `###` heading), each one a complete paragraph on one physical line.

**`.claude/skills/demo-reel-adapter/SKILL.md:237-243`** — the flagged lines are **numbered list item 4** of the acceptance-test list, and CLAUDE.md rule 10 exempts exactly that: *"This applies only to prose: keep the natural line structure of code blocks, tables, **list items**, frontmatter, and intentional hard breaks."* Items 1-3 immediately above it are wrapped the same way and predate this PR, so unwrapping item 4 alone would make the list inconsistent, and unwrapping all four would be unrelated churn in a PR about cast provenance.

Every genuine prose paragraph this PR adds to `SKILL.md` is single-line — the two short ones (172 and 178 characters) are short paragraphs, not wrapped ones. Verified by filtering the added lines to non-list, non-table, non-code content and checking there is no sub-200-character continuation line.

Happy to be shown a specific wrapped prose paragraph I have missed.

[the remark this answers is not in the document]
Comment (prageethw) [reply .claude/skills/demo-reel-adapter/SKILL.md:243]: Confirmed independently — the review is accurate. Resolving.

**Review by prageethw — APPROVED** [review]
Reviewed the full diff (adapter shell logic, Rust harness `write_provenance`/`recording_build_commit`, and both test suites). The provenance gate is fail-closed throughout (absent/corrupt/wrong-schema/wrong-commit/non-passing sidecars all refuse; no commit to check against `die`s rather than publishing unchecked), per-clip rather than all-or-nothing, and the harness/adapter field contract is cross-checked by `the_provenance_contract_the_adapter_reads_is_the_one_the_harness_writes`. `write_provenance` also runs its output through the redactor even though no field carries process output today — good defense in depth. The one open thread (hard-wrap review) was a bot false-positive the author correctly rebutted with measurements and the bot conceded; resolved. `e2e-deterministic` failure is documented pre-existing/non-required (fixed by #837). LGTM.
